cpp/src/arrow/flight/sql/odbc/odbc_impl/odbc_handle.h

```diff
@@ -47,7 +47,18 @@ class ODBCHandle {
     try {
       GetDiagnostics().Clear();
       rc = function();
-    } catch (const arrow::flight::sql::odbc::DriverException& ex) {
+    } catch (const arrow::flight::sql::odbc::AuthenticationException& ex) {
+      GetDiagnostics().AddError(arrow::flight::sql::odbc::DriverException(
+          ex.GetMessageText(), ex.GetSqlState(), ex.GetNativeError()));
+    } catch (const arrow::flight::sql::odbc::NullWithoutIndicatorException& ex) {
+      GetDiagnostics().AddError(arrow::flight::sql::odbc::DriverException(
+          ex.GetMessageText(), ex.GetSqlState(), ex.GetNativeError()));
+    }
+    // on mac, DriverException doesn't catch the subclass exceptions hence we added
+    // the following above.
+    // TODO investigate if there is a way to catch all the subclass exceptions under
```

Comment:
```suggestion
    // GH-48278 TODO investigate if there is a way to catch all the subclass exceptions under
```
(might need lint fixes)
https://github.com/apache/arrow/issues/48278